CA1816 shouldn't warn for IAsyncDisposable.DisposeAsync

#### Analyzer package

[Microsoft.NetCore.Analyzers](https://www.nuget.org/packages/Microsoft.NetCore.Analyzers/)

#### Package Version

v3.0.0 (Latest)

#### Diagnostic ID

[CA1816](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1816)

#### Repro steps

Create a class which implements [`IAsyncDisposable`](https://docs.microsoft.com/en-us/dotnet/api/system.iasyncdisposable) and calls [`GC.SuppressFinalize`](https://docs.microsoft.com/en-us/dotnet/api/system.gc.suppressfinalize) in [`DisposeAsync`](https://docs.microsoft.com/en-us/dotnet/api/system.iasyncdisposable.disposeasync).  For example:

```csharp
namespace CA1816Example
{
    using System;
    using System.Runtime.InteropServices;
    using System.Threading.Tasks;

    class MyAsyncDisposable : IAsyncDisposable
    {
        [DllImport("example.dll")]
        private static extern int GetHandle();

        [DllImport("example.dll")]
        private static extern void FreeHandle(int handle);

        private readonly int handle;

        public MyAsyncDisposable()
        {
            this.handle = GetHandle();
        }

        ~MyAsyncDisposable()
        {
            FreeHandle(this.handle);
        }

        public async ValueTask DisposeAsync()
        {
            await Task.Run(() => FreeHandle(this.handle)).ConfigureAwait(false);
            GC.SuppressFinalize(this);
        }
    }
}
```

#### Expected behavior

CA1816 would not be produced.

#### Actual behavior

    MyAsyncDisposable.cs(30,13): warning CA1816: MyAsyncDisposable.DisposeAsync() calls GC.SuppressFinalize(object), a method that is typically only called within an implementation of 'IDisposable.Dispose'. Refer to the IDisposable pattern for more information.

#### Discussion

Although I haven't seen official guidance for [dispose pattern](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#implement-the-dispose-pattern) with `IAsyncDisposable`, it seems likely that classes which define a finalizer (e.g. to release native resources) would want to suppress finalization if the work has been completed by `DisposeAsync`, as in the above example, and therefore CA1816 shouldn't apply to use in this method.  Thoughts?

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectly.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectly.cs
@@ -183,7 +183,7 @@ namespace Microsoft.NetCore.Analyzers.Runtime
                     return SuppressFinalizeUsage.CanCall;
                 }
 
-                if (!method.IsDisposeImplementation(_compilation))
+                if (!method.IsDisposeImplementation(_compilation) && !method.IsAsyncDisposeImplementation(_compilation))
                 {
                     return SuppressFinalizeUsage.MustNotCall;
                 }

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
@@ -181,6 +181,7 @@ class MyAsyncDisposable : IAsyncDisposable
         public async Task AsyncDisposableWithFinalizer_Basic_NoDiagnostic()
         {
             var code = @"
+Imports System
 Imports System.Runtime.InteropServices
 Imports System.Threading.Tasks
 

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
@@ -178,6 +178,44 @@ class MyAsyncDisposable : IAsyncDisposable
         }
 
         [Fact]
+        public async Task AsyncDisposableWithFinalizer_Basic_NoDiagnostic()
+        {
+            var code = @"
+Imports System.Runtime.InteropServices
+Imports System.Threading.Tasks
+
+Class MyAsyncDisposable
+    Inherits IAsyncDisposable
+
+    <DllImport(""example.dll"")>
+    Private Shared Function GetHandle() As Integer
+
+    <DllImport(""example.dll"")>
+    Private Shared Sub FreeHandle(handle As Integer)
+
+    Private ReadOnly handle As Integer
+
+    Public Sub New()
+        Me.handle = GetHandle()
+    End Sub
+
+    Protected Overrides Sub Finalize()
+        FreeHandle(Me.handle)
+    End Sub
+
+    Public Async Function DisposeAsync() As ValueTask
+        Await Task.Run(Sub() FreeHandle(Me.handle)).ConfigureAwait(False)
+        GC.SuppressFinalize(Me)
+    End Function
+End Class";
+            await new VerifyVB.Test
+            {
+                ReferenceAssemblies = AdditionalMetadataReferences.DefaultWithAsyncInterfaces,
+                TestCode = code
+            }.RunAsync();
+        }
+        
+        [Fact]
         public async Task SealedDisposableWithoutFinalizer_CSharp_NoDiagnostic()
         {
 

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
@@ -3,6 +3,7 @@
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Testing;
+using Test.Utilities;
 using Xunit;
 using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
     Microsoft.NetCore.Analyzers.Runtime.CallGCSuppressFinalizeCorrectlyAnalyzer,

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
@@ -169,7 +169,11 @@ class MyAsyncDisposable : IAsyncDisposable
         GC.SuppressFinalize(this);
     }
 }";
-            await VerifyCS.VerifyAnalyzerAsync(code);
+            await new VerifyCS.Test
+            {
+                ReferenceAssemblies = AdditionalMetadataReferences.DefaultWithAsyncInterfaces,
+                TestCode = code
+            }.RunAsync();
         }
 
         [Fact]

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
@@ -136,6 +136,43 @@ End Class";
         }
 
         [Fact]
+        public async Task AsyncDisposableWithFinalizer_CSharp_NoDiagnostic()
+        {
+            var code = @"
+using System;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+class MyAsyncDisposable : IAsyncDisposable
+{
+    [DllImport(""example.dll"")]
+    private static extern int GetHandle();
+
+    [DllImport(""example.dll"")]
+    private static extern void FreeHandle(int handle);
+
+    private readonly int handle;
+
+    public MyAsyncDisposable()
+    {
+        this.handle = GetHandle();
+    }
+
+    ~MyAsyncDisposable()
+    {
+        FreeHandle(this.handle);
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        await Task.Run(() => FreeHandle(this.handle)).ConfigureAwait(false);
+        GC.SuppressFinalize(this);
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(code);
+        }
+
+        [Fact]
         public async Task SealedDisposableWithoutFinalizer_CSharp_NoDiagnostic()
         {
 

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
@@ -185,13 +185,15 @@ Imports System.Runtime.InteropServices
 Imports System.Threading.Tasks
 
 Class MyAsyncDisposable
-    Inherits IAsyncDisposable
+    Implements IAsyncDisposable
 
     <DllImport(""example.dll"")>
     Private Shared Function GetHandle() As Integer
+    End Function
 
     <DllImport(""example.dll"")>
     Private Shared Sub FreeHandle(handle As Integer)
+    End Sub
 
     Private ReadOnly handle As Integer
 
@@ -214,7 +216,7 @@ End Class";
                 TestCode = code
             }.RunAsync();
         }
-        
+
         [Fact]
         public async Task SealedDisposableWithoutFinalizer_CSharp_NoDiagnostic()
         {

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/CallGCSuppressFinalizeCorrectlyTests.cs
@@ -178,47 +178,6 @@ class MyAsyncDisposable : IAsyncDisposable
         }
 
         [Fact]
-        public async Task AsyncDisposableWithFinalizer_Basic_NoDiagnostic()
-        {
-            var code = @"
-Imports System
-Imports System.Runtime.InteropServices
-Imports System.Threading.Tasks
-
-Class MyAsyncDisposable
-    Implements IAsyncDisposable
-
-    <DllImport(""example.dll"")>
-    Private Shared Function GetHandle() As Integer
-    End Function
-
-    <DllImport(""example.dll"")>
-    Private Shared Sub FreeHandle(handle As Integer)
-    End Sub
-
-    Private ReadOnly handle As Integer
-
-    Public Sub New()
-        Me.handle = GetHandle()
-    End Sub
-
-    Protected Overrides Sub Finalize()
-        FreeHandle(Me.handle)
-    End Sub
-
-    Public Async Function DisposeAsync() As ValueTask
-        Await Task.Run(Sub() FreeHandle(Me.handle)).ConfigureAwait(False)
-        GC.SuppressFinalize(Me)
-    End Function
-End Class";
-            await new VerifyVB.Test
-            {
-                ReferenceAssemblies = AdditionalMetadataReferences.DefaultWithAsyncInterfaces,
-                TestCode = code
-            }.RunAsync();
-        }
-
-        [Fact]
         public async Task SealedDisposableWithoutFinalizer_CSharp_NoDiagnostic()
         {
 

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -1,13 +1,13 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Collections.Immutable;
-using System.Linq;
-using Microsoft.CodeAnalysis;
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
-using System.Collections.Concurrent;
+using System.Linq;
+using Microsoft.CodeAnalysis;
 
 #if HAS_IOPERATION
 using System.Threading;
@@ -194,7 +194,7 @@ namespace Analyzer.Utilities.Extensions
         }
 
         /// <summary>
-        /// Checks if the given method implements <see cref="IAsyncDisposable.Dispose"/> or overrides an implementation of <see cref="IAsyncDisposable.Dispose"/>.
+        /// Checks if the given method implements "IAsyncDisposable.Dispose" or overrides an implementation of "IAsyncDisposable.Dispose".
         /// </summary>
         public static bool IsAsyncDisposeImplementation([NotNullWhen(returnValue: true)] this IMethodSymbol? method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? iAsyncDisposable)
         {

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -212,7 +212,7 @@ namespace Analyzer.Utilities.Extensions
             // if it is the given method.
             return method.ReturnsVoid &&
                 method.Parameters.Length == 0 &&
-                method.IsImplementationOfInterfaceMethod(null, iAsyncDisposable, "Dispose");
+                method.IsImplementationOfInterfaceMethod(null, iAsyncDisposable, "DisposeAsync");
         }
 
         /// <summary>

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -168,7 +168,7 @@ namespace Analyzer.Utilities.Extensions
         public static bool IsAsyncDisposeImplementation(this IMethodSymbol method, Compilation compilation)
         {
             INamedTypeSymbol? iAsyncDisposable = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIAsyncDisposable);
-            return method.IsDisposeImplementation(iAsyncDisposable);
+            return method.IsAsyncDisposeImplementation(iAsyncDisposable);
         }
 
         /// <summary>
@@ -191,6 +191,28 @@ namespace Analyzer.Utilities.Extensions
             return method.ReturnsVoid &&
                 method.Parameters.Length == 0 &&
                 method.IsImplementationOfInterfaceMethod(null, iDisposable, "Dispose");
+        }
+
+        /// <summary>
+        /// Checks if the given method implements <see cref="IAsyncDisposable.Dispose"/> or overrides an implementation of <see cref="IAsyncDisposable.Dispose"/>.
+        /// </summary>
+        public static bool IsAsyncDisposeImplementation([NotNullWhen(returnValue: true)] this IMethodSymbol? method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? iAsyncDisposable)
+        {
+            if (method == null)
+            {
+                return false;
+            }
+
+            if (method.IsOverride)
+            {
+                return method.OverriddenMethod.IsAsyncDisposeImplementation(iAsyncDisposable);
+            }
+
+            // Identify the implementor of IAsyncDisposable.Dispose in the given method's containing type and check
+            // if it is the given method.
+            return method.ReturnsVoid &&
+                method.Parameters.Length == 0 &&
+                method.IsImplementationOfInterfaceMethod(null, iAsyncDisposable, "Dispose");
         }
 
         /// <summary>

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -211,7 +211,7 @@ namespace Analyzer.Utilities.Extensions
 
             // Identify the implementor of IAsyncDisposable.Dispose in the given method's containing type and check
             // if it is the given method.
-            return method.ReturnType != null && method.ReturnType.Equals(valueTaskType) && method.Parameters.Length == 0 &&
+            return method.ReturnType != null && method.ReturnType.Equals(valueTaskType) && method.Parameters.IsEmpty &&
                 method.IsImplementationOfInterfaceMethod(null, iAsyncDisposable, "DisposeAsync");
         }
 

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -208,9 +208,10 @@ namespace Analyzer.Utilities.Extensions
                 return method.OverriddenMethod.IsAsyncDisposeImplementation(iAsyncDisposable);
             }
 
+            INamedTypeSymbol? valueTaskType= compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemValueTask);
             // Identify the implementor of IAsyncDisposable.Dispose in the given method's containing type and check
             // if it is the given method.
-            return method.Parameters.Length == 0 &&
+            return method.ReturnType.Equals(valueTaskType) && method.Parameters.Length == 0 &&
                 method.IsImplementationOfInterfaceMethod(null, iAsyncDisposable, "DisposeAsync");
         }
 

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -210,8 +210,7 @@ namespace Analyzer.Utilities.Extensions
 
             // Identify the implementor of IAsyncDisposable.Dispose in the given method's containing type and check
             // if it is the given method.
-            return Equals(method.ReturnType, iAsyncDisposable) &&
-                method.Parameters.Length == 0 &&
+            return method.Parameters.Length == 0 &&
                 method.IsImplementationOfInterfaceMethod(null, iAsyncDisposable, "DisposeAsync");
         }
 

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -210,7 +210,7 @@ namespace Analyzer.Utilities.Extensions
 
             // Identify the implementor of IAsyncDisposable.Dispose in the given method's containing type and check
             // if it is the given method.
-            return method.ReturnsVoid &&
+            return Equals(method.ReturnType, iAsyncDisposable) &&
                 method.Parameters.Length == 0 &&
                 method.IsImplementationOfInterfaceMethod(null, iAsyncDisposable, "DisposeAsync");
         }

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -159,8 +159,16 @@ namespace Analyzer.Utilities.Extensions
         public static bool IsDisposeImplementation(this IMethodSymbol method, Compilation compilation)
         {
             INamedTypeSymbol? iDisposable = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIDisposable);
+            return method.IsDisposeImplementation(iDisposable);
+        }
+
+        /// <summary>
+        /// Checks if the given method implements IAsyncDisposable.Dispose()
+        /// </summary>
+        public static bool IsAsyncDisposeImplementation(this IMethodSymbol method, Compilation compilation)
+        {
             INamedTypeSymbol? iAsyncDisposable = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIAsyncDisposable);
-            return method.IsDisposeImplementation(iDisposable) || method.IsDisposeImplementation(iAsyncDisposable);
+            return method.IsDisposeImplementation(iAsyncDisposable);
         }
 
         /// <summary>

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -168,7 +168,8 @@ namespace Analyzer.Utilities.Extensions
         public static bool IsAsyncDisposeImplementation(this IMethodSymbol method, Compilation compilation)
         {
             INamedTypeSymbol? iAsyncDisposable = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIAsyncDisposable);
-            return method.IsAsyncDisposeImplementation(iAsyncDisposable);
+            INamedTypeSymbol? valueTaskType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemValueTask);
+            return method.IsAsyncDisposeImplementation(iAsyncDisposable, valueTaskType);
         }
 
         /// <summary>
@@ -196,7 +197,7 @@ namespace Analyzer.Utilities.Extensions
         /// <summary>
         /// Checks if the given method implements "IAsyncDisposable.Dispose" or overrides an implementation of "IAsyncDisposable.Dispose".
         /// </summary>
-        public static bool IsAsyncDisposeImplementation([NotNullWhen(returnValue: true)] this IMethodSymbol? method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? iAsyncDisposable)
+        public static bool IsAsyncDisposeImplementation([NotNullWhen(returnValue: true)] this IMethodSymbol? method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? iAsyncDisposable, INamedTypeSymbol? valueTaskType)
         {
             if (method == null)
             {
@@ -205,10 +206,9 @@ namespace Analyzer.Utilities.Extensions
 
             if (method.IsOverride)
             {
-                return method.OverriddenMethod.IsAsyncDisposeImplementation(iAsyncDisposable);
+                return method.OverriddenMethod.IsAsyncDisposeImplementation(iAsyncDisposable, valueTaskType);
             }
 
-            INamedTypeSymbol? valueTaskType= compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemValueTask);
             // Identify the implementor of IAsyncDisposable.Dispose in the given method's containing type and check
             // if it is the given method.
             return method.ReturnType.Equals(valueTaskType) && method.Parameters.Length == 0 &&

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -197,7 +197,7 @@ namespace Analyzer.Utilities.Extensions
         /// <summary>
         /// Checks if the given method implements "IAsyncDisposable.Dispose" or overrides an implementation of "IAsyncDisposable.Dispose".
         /// </summary>
-        public static bool IsAsyncDisposeImplementation([NotNullWhen(returnValue: true)] this IMethodSymbol? method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? iAsyncDisposable, INamedTypeSymbol? valueTaskType)
+        public static bool IsAsyncDisposeImplementation([NotNullWhen(returnValue: true)] this IMethodSymbol? method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? iAsyncDisposable, [NotNullWhen(returnValue: true)] INamedTypeSymbol? valueTaskType)
         {
             if (method == null)
             {
@@ -211,7 +211,7 @@ namespace Analyzer.Utilities.Extensions
 
             // Identify the implementor of IAsyncDisposable.Dispose in the given method's containing type and check
             // if it is the given method.
-            return method.ReturnType.Equals(valueTaskType) && method.Parameters.Length == 0 &&
+            return method.ReturnType != null && method.ReturnType.Equals(valueTaskType) && method.Parameters.Length == 0 &&
                 method.IsImplementationOfInterfaceMethod(null, iAsyncDisposable, "DisposeAsync");
         }
 

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -159,7 +159,8 @@ namespace Analyzer.Utilities.Extensions
         public static bool IsDisposeImplementation(this IMethodSymbol method, Compilation compilation)
         {
             INamedTypeSymbol? iDisposable = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIDisposable);
-            return method.IsDisposeImplementation(iDisposable);
+            INamedTypeSymbol? iAsyncDisposable = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIAsyncDisposable);
+            return method.IsDisposeImplementation(iDisposable) || method.IsDisposeImplementation(iAsyncDisposable);
         }
 
         /// <summary>

--- a/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
+++ b/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs
@@ -168,7 +168,7 @@ namespace Analyzer.Utilities.Extensions
         public static bool IsAsyncDisposeImplementation(this IMethodSymbol method, Compilation compilation)
         {
             INamedTypeSymbol? iAsyncDisposable = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIAsyncDisposable);
-            INamedTypeSymbol? valueTaskType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemValueTask);
+            INamedTypeSymbol? valueTaskType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksValueTask);
             return method.IsAsyncDisposeImplementation(iAsyncDisposable, valueTaskType);
         }
 

